shelley/chain-and-ledger/executable-spec/test/Generator/Utxo.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- The HasTrace instance relies on test generators and
+-- hence not situated with the LEDGER STS.
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Generator.Utxo
+  ( genTx
+  )
+  where
+
+import           Control.Monad (when)
+import qualified Data.Map.Strict as Map
+import           Data.Sequence (Seq)
+import           Data.Set (Set)
+import           Hedgehog (Gen)
+
+import qualified Hedgehog.Gen as Gen
+
+import           Cardano.Crypto.DSIGN.Mock (MockDSIGN)
+import           Cardano.Crypto.Hash (ShortHash)
+
+import           Coin (Coin (..))
+import           Generator.Core (findPayKeyPair, genInteger, genNatural, toAddr)
+import           Generator.Delegation (genDCerts)
+import           Ledger.Core (dom)
+import           LedgerState (pattern UTxOState)
+import           MockTypes (Addr, DCert, DPState, KeyPair, KeyPairs, Tx, TxBody, TxIn, TxOut, UTxO,
+                     UTxOState)
+import           Slot (Slot (..))
+import           STS.Ledger (LedgerEnv (..))
+import           Tx (pattern Tx, pattern TxBody, pattern TxOut)
+import           Updates (emptyUpdate)
+import           UTxO (pattern UTxO, balance, makeWitnessesVKey)
+
+-- | Generate a new transaction in the context of the LEDGER STS environment and state.
+--
+-- Selects unspent outputs and spends the funds on a some valid addresses.
+-- Also generates valid certificates.
+genTx :: LedgerEnv ShortHash MockDSIGN
+      -> (UTxOState, DPState)
+      -> KeyPairs
+      -> Gen Tx
+genTx (LedgerEnv _ slot _ pparams) (UTxOState utxo _ _ _, dpState) keys = do
+  keys' <- Gen.shuffle keys
+
+  -- inputs
+  (inputs, spendingBalance, spendWits) <- pickSpendingInputs keys' utxo
+
+  -- output addresses
+  recipientAddrs <- genRecipients keys'
+
+  -- certificates
+  (certs, certWits, deposits_) <- genDCerts keys' pparams dpState
+
+  -- attempt to make provision for certificate deposits (otherwise discard this generator)
+  when (spendingBalance < deposits_) Gen.discard
```

Comment:
How often does this `discard` occur? I mean what's the probability of this occurring? I'm worried this might be overly inefficient.

Reply:
As it stands it's very rare, I'll keep an eye on it.